[동기] Step3,4 : HTTP 웹서버 구현

donggi-lee-bit opened the pull request:
### Description

요구사항 3 : POST 방식으로 회원가입하기까지 진행하였습니다.
기능 구현한 뒤에 리팩토링을 하고자 했습니다.
과제 초반엔 기능 구현을 하고 테스트를 통과하고 다음 요구사항으로 넘어가자는 마음이었는데
갈수록 시간에 쫓기게 되니 테스트를 제대로 하지 못하게 되었습니다.

### 학습 내용

[notion에 정리한 학습 내용](https://victorious-lint-a8e.notion.site/next-step-e74a505d0cb04bd5a6f08b922411ba30)

### 결과

ex) 실행, 테스트 결과 또는 사진 첨부

<img width="517" alt="Screen Shot 2022-07-23 at 8 41 04 PM" src="https://user-images.githubusercontent.com/73376468/180603439-d237a051-289c-4b22-9f4a-b81ace1e90f1.png">

donggi-lee-bit committed:
chore: gitignore 파일 추가

donggi-lee-bit committed:
feat: 책 Calculator 예제

donggi-lee-bit committed:
feat: StringCalculator 구현

donggi-lee-bit committed:
Merge pull request #2 from donggi-lee-bit/step2/calculator

chapter2 요구사항 구현

donggi-lee-bit committed:
refactor: add() 메서드가 가진 여러 역할 리팩토링

- 구분자를 찾는 메서드
- int로 파싱하는 메서드
- null 혹은 빈 문자열을 받는 메서드

로 메서드가 한 가지 일만 하게끔 리팩토링

donggi-lee-bit committed:
Merge pull request #4 from donggi-lee-bit/ch2/refactor

add() 메서드가 가진 여러 역할 리팩토링

donggi-lee-bit committed:
feat: index.html 응답 구현

- 먼저 request line을 입력 받는다
- Files.readAllBytes() 메서드를 통해 ./webapp 경로에 있는 index.html 파일에 접근하여 읽어올 수 있었다

donggi-lee-bit committed:
Merge pull request #6 from donggi-lee-bit/ch3/get-index.html

feat: index.html 응답 구현

donggi-lee-bit committed:
feat: GET 방식으로 회원가입하기

donggi-lee-bit committed:
Merge pull request #8 from donggi-lee-bit/ch3/get-index.html

GET 방식으로 회원가입하기

donggi-lee-bit committed:
fix: 유저 생성 로직 에러 수정

- 기존 파싱에 문제가 있어 생긴 문제 책의 코드로 수정

donggi-lee-bit committed:
feat: POST 요청으로 회원가입 기능 구현

- decode 해주는 메서드 deprecated 되어 수정해야함
- header 의 content-length를 읽어 IOUtils.readData() 메서드를 통해 body에 담아져 온 데이터를 변환하는 과정이 있었다.

donggi-lee-bit committed:
Merge pull request #10 from donggi-lee-bit/ch3/feat-post-login

POST 요청으로 회원가입 기능 구현

donggi-lee-bit committed:
feat: 302 Redirect 구현하기

- POST 요청으로 회원가입을 할 때 User를 만든 뒤 302 response 메서드를 호출하여 /index.html 로 redirect하도록 구현

donggi-lee-bit committed:
Merge pull request #12 from donggi-lee-bit/ch3/feat-302statuscode

302 Redirect 구현하기

ku-kim commented on src/test/java/calculator/CalculatorTest.java:
```diff
@@ -0,0 +1,37 @@
+        System.out.println("before");
+    }
+
+    @Test
+    void add() {
+        int result = cal.add(6, 3);
+        assertThat(result).isEqualTo(9);
+    }
```
문자열 계산기 구현 코드도 함께 들어있군요! 
좋네요 ㅎㅎ

ku-kim commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,54 @@
+        cal = new StringCalculator();
+    }
+
+    @Test
+    void add() {
+        String text = "1,2,3";
+        int result = cal.add(text);
+        assertThat(result).isEqualTo(6);
+    }
+
+    @Test
+    void add2() {
+        String text = "1:2,3";
+        int result = cal.add(text);
+        assertThat(result).isEqualTo(6);
+    }
```
given, when, then (또는 AAA)패턴으로 작성하신건가요?

주석이나 개행으로 구분하면 테스트코드 가독성이 높아질 거 같아요!
개인적으로는 개행으로 구분하고 있어요!

ku-kim commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,54 @@
+        assertThat(result).isEqualTo(6);
+    }
+
+    @Test
+    void add2() {
+        String text = "1:2,3";
+        int result = cal.add(text);
+        assertThat(result).isEqualTo(6);
+    }
```
given, when, then (또는 AAA)패턴으로 작성하신건가요?

주석이나 개행으로 구분하면 테스트코드 가독성이 높아질 거 같은데 어떠신가요?
개인적으로 개행으로 구분하고 있어요!

### 
```suggestion
    @Test
    void add2() {
        String text = "1:2,3";
        
        int result = cal.add(text);
        
        assertThat(result).isEqualTo(6);
    }
```

donggi-lee-bit replied:
restAssured로 작성할 땐 given when then을 명시할 수 있어서 편했는데
이러한 테스트에서는 어떻게 구분하고 있는지 몰랐네요! 개행으로 구분해보도록 하겠습니다. 꿀팁 감사해요!

ku-kim commented on src/test/java/calculator/StringCalculatorTest.java:
```diff
@@ -0,0 +1,54 @@
+        assertThrows(RuntimeException.class, () -> cal.add(text));
+    }
+
+    @Test
+    void emptyString() {
+        String text = null;
+        String text2 = "";
+        int result = cal.add(text);
+        int result2 = cal.add(text2);
+        assertThat(result).isZero();
+        assertThat(result2).isZero();
+    }
```
```suggestion
    @ParameterizedTest
    @NullAndEmptySource
    void invalidString(string text) {
        int result = cal.add(text);
        
        assertThat(result).isZero();
    }
```

emptyString() 의 경우 하나의 테스트 코드에 두 가지 경우(null, "") 를 테스트하고 있네요.

null과 "" 은 다른 경우라고 생각되는데요.
nullString(), emptyString()으로 테스트 코드를 두 번 작성하거나
편의성을 위해 @ParameterizedTest 를 사용해보는건 어떠신가요? 

https://www.baeldung.com/parameterized-tests-junit-5

donggi-lee-bit replied:
@ParameterizedTest 가 있었네요! 그리고 말씀하신 것처럼 null과 빈 문자열은 확실히 다른 것 같습니다 ㅎㅎ;;; 예시까지 들어주신@ParameterizedTest 사용을 해보도록 하겠습니다! 감사해용 ㅎㅎ

ku-kim reviewed (approved):
테스트코드 도전 좋은거 같아요!
앞으로의 step에서 코드 기대할게요 동기!

donggi-lee-bit committed:
feat: 요구사항에 맞춰 로그인 기능 구현

- 입력 받은 userId와 password를 DataBase에서 찾은 userId을 찾아와서 password를 확인하여 login 성공 실패 여부에 맞게 응답하게 구현

donggi-lee-bit committed:
refactor: LoginService 싱글톤으로 동작하도록 변경
